NBitcoin/Protocol/NetworkAddress.cs

```diff
@@ -120,19 +428,170 @@ public void ReadWrite(BitcoinStream stream)
 
 		#endregion
 
-		public void ZeroTime()
+		private BIP155Network GetBIP155Network()
 		{
-			this.ntime = 0;
+			switch ((Network)network) 
+			{
+				case Network.IPv4:
+					return BIP155Network.IPv4;
+				case Network.IPv6:
+					return BIP155Network.IPv6;
+				case Network.Onion:
+					if (addr.Length == ADDR_TORV2_SIZE)
+					{
+						return BIP155Network.TORv2;
+					}
+					else if (addr.Length == ADDR_TORV3_SIZE)
+					{
+						return BIP155Network.TORv3;
+					}
+					throw new Exception("Onion address size doesn't match the TORv2 nor TORv3 sizes.");
+				case Network.I2P:
+					return BIP155Network.I2P;
+				case Network.Cjdns:
+					return BIP155Network.Cjdns;
+				case Network.Unroutable: // m_net is never and should not be set to NET_UNROUTABLE
+				case Network.Max:        // m_net is never and should not be set to NET_MAX
+					throw new Exception("Unknown network.");
+			} // no default case, so the compiler can warn about missing cases
+			throw new Exception("Unknown network.");
 		}
 
-		internal byte[] GetKey()
+		private bool SetNetFromBIP155Network(BIP155Network bip155Network, int length)
 		{
-			var vKey = new byte[18];
-			Array.Copy(ip, vKey, 16);
-			vKey[16] = (byte)(port / 0x100);
-			vKey[17] = (byte)(port & 0x0FF);
+			byte AssignIfCorrectSize(Network net, int expectedSize)
+			{
+				if (length != expectedSize)
+				{
+					throw new Exception($"BIP155 {(BIP155Network)bip155Network} address with length {length} (should be {expectedSize}).");

```

Comment:
Why is that `(BIP155Network)` cast needed?

```suggestion
					throw new ArgumentException($"Provided BIP155 {bip155Network} address is {length} bytes long. Expected length is {expectedSize} bytes.");
```

Maybe something like this?

Reply:
I take it. The cast is needed because we want to see the IPv4 instead of 0, for example.